File: telephony/livekit_room.go
package main

import (
	"context"
	"fmt"
	"log/slog"
	"strings"
	"time"

	"github.com/livekit/protocol/livekit"
	lksdk "github.com/livekit/server-sdk-go/v2"
)

func liveKitRoomName(callerNumber string) string {
	prefix := "voice-"
	if cfg != nil && cfg.LiveKitRoomPrefix != "" {
		prefix = cfg.LiveKitRoomPrefix
	}
	return prefix + callerNumber
}

func liveKitRoomServiceURL(liveKitURL string) string {
	s := strings.TrimSpace(liveKitURL)
	s = strings.TrimSuffix(s, "/")
	if strings.HasPrefix(s, "wss://") {
		return "https://" + strings.TrimPrefix(s, "wss://")
	}
	if strings.HasPrefix(s, "ws://") {
		return "http://" + strings.TrimPrefix(s, "ws://")
	}
	return s
}

func sipUserPart(sipURI string) string {
	s := strings.TrimSpace(sipURI)
	s = strings.TrimPrefix(strings.ToLower(s), "sip:")
	if at := strings.Index(s, "@"); at >= 0 {
		return s[:at]
	}
	return ""
}

func waitForDispatchedLiveKitRoom(ctx context.Context, callerNumber string, sipCallee string) (string, error) {
	if cfg.LiveKitAPIKey == "" || cfg.LiveKitAPISecret == "" || cfg.LiveKitURL == "" {
		return "", fmt.Errorf("LiveKit credentials not configured")
	}

	roomSvc := lksdk.NewRoomServiceClient(liveKitRoomServiceURL(cfg.LiveKitURL), cfg.LiveKitAPIKey, cfg.LiveKitAPISecret)
	prefix := cfg.LiveKitRoomPrefix
	if prefix == "" {
		prefix = "voice-"
	}

	var candidatePrefixes []string
	if callerNumber != "" {
		candidatePrefixes = append(candidatePrefixes, prefix+callerNumber)
	}
	if sipCallee != "" {
		candidatePrefixes = append(candidatePrefixes, prefix+sipCallee)
	}
	candidatePrefixes = append(candidatePrefixes, prefix)

	ticker := time.NewTicker(750 * time.Millisecond)
	defer ticker.Stop()

	deadline := time.NewTimer(25 * time.Second)
	defer deadline.Stop()

	for {
		roomsResp, err := roomSvc.ListRooms(ctx, &livekit.ListRoomsRequest{})
		if err == nil {
			for _, room := range roomsResp.Rooms {
				if room == nil {
					continue
				}
				matches := false
				for _, pfx := range candidatePrefixes {
					if strings.HasPrefix(room.Name, pfx) {
						matches = true
						break
					}
				}
				if !matches {
					continue
				}

				partsResp, pErr := roomSvc.ListParticipants(ctx, &livekit.ListParticipantsRequest{Room: room.Name})
				if pErr != nil {
					continue
				}
				for _, p := range partsResp.Participants {
					if p == nil {
						continue
					}
					id := p.Identity
					if strings.HasPrefix(id, "agent-") || strings.HasPrefix(id, "system-hold-") {
						continue
					}
					if strings.HasPrefix(id, "sip_") || (callerNumber != "" && strings.Contains(id, callerNumber)) || (sipCallee != "" && strings.Contains(id, sipCallee)) {
						return room.Name, nil
					}
				}
			}
		} else {
			slog.Warn("Failed to list LiveKit rooms while waiting for SIP dispatch", "err", err)
		}

		select {
		case <-ctx.Done():
			return "", ctx.Err()
		case <-deadline.C:
			return "", fmt.Errorf("timed out waiting for LiveKit SIP dispatch room")
		case <-ticker.C:
		}
	}
}

File: telephony/src/lib/livekitRoom.ts
// telephony/src/lib/livekitRoom.ts
// Converted from livekit_room.go

import { RoomServiceClient } from "livekit-server-sdk";
import { getConfig } from "./config";

export function liveKitRoomName(callerNumber: string): string {
  const cfg = getConfig();
  const prefix = cfg.liveKitRoomPrefix || "voice-";
  return prefix + callerNumber;
}

function liveKitRoomServiceURL(liveKitURL: string): string {
  let s = liveKitURL.trim().replace(/\/$/, "");
  if (s.startsWith("wss://")) return "https://" + s.slice("wss://".length);
  if (s.startsWith("ws://")) return "http://" + s.slice("ws://".length);
  return s;
}

export function sipUserPart(sipURI: string): string {
  let s = sipURI.trim().replace(/^sip:/i, "");
  const at = s.indexOf("@");
  return at >= 0 ? s.slice(0, at) : "";
}

export async function waitForDispatchedLiveKitRoom(
  signal: AbortSignal,
  callerNumber: string,
  sipCallee: string
): Promise<string> {
  const cfg = getConfig();
  if (!cfg.liveKitAPIKey || !cfg.liveKitAPISecret || !cfg.liveKitURL) {
    throw new Error("LiveKit credentials not configured");
  }

  const roomSvc = new RoomServiceClient(
    liveKitRoomServiceURL(cfg.liveKitURL),
    cfg.liveKitAPIKey,
    cfg.liveKitAPISecret
  );

  const prefix = cfg.liveKitRoomPrefix || "voice-";
  const candidatePrefixes: string[] = [];
  if (callerNumber) candidatePrefixes.push(prefix + callerNumber);
  if (sipCallee) candidatePrefixes.push(prefix + sipCallee);
  candidatePrefixes.push(prefix);

  const deadline = Date.now() + 25_000;

  while (Date.now() < deadline && !signal.aborted) {
    try {
      const rooms = await roomSvc.listRooms();
      for (const room of rooms) {
        if (!room.name) continue;
        const matches = candidatePrefixes.some((pfx) => room.name.startsWith(pfx));
        if (!matches) continue;

        const participants = await roomSvc.listParticipants(room.name);
        for (const p of participants) {
          const id = p.identity ?? "";
          if (id.startsWith("agent-") || id.startsWith("system-hold-")) continue;
          if (
            id.startsWith("sip_") ||
            (callerNumber && id.includes(callerNumber)) ||
            (sipCallee && id.includes(sipCallee))
          ) {
            return room.name;
          }
        }
      }
    } catch (err) {
      console.warn("[livekit] Failed to list rooms while waiting for SIP dispatch", err);
    }

    await new Promise<void>((resolve, reject) => {
      if (signal.aborted) return reject(new Error("aborted"));
      const t = setTimeout(resolve, 750);
      signal.addEventListener("abort", () => { clearTimeout(t); reject(new Error("aborted")); });
    });
  }

  throw new Error("Timed out waiting for LiveKit SIP dispatch room");
}
